src/os/bluestore/BlueStore.cc

```diff
@@ -9195,12 +9332,44 @@ int BlueStore::_fsck(BlueStore::FSCKDepth depth, bool repair)
       depth == FSCK_SHALLOW ? " (shallow)" : " (regular)")
     << dendl;
 
+  string p = path + "/block";
+  bluestore_bdev_label_t label;
+  if (!repair) {
+    int r = _read_bdev_label(cct, p, &label, false);
+    if (r < 0) {
+      if (r == -EIO) {
+        derr << __func__ << " fsck error: corrupted block device labels found" << dendl;
+      }
+      return r;
+    }
+  } else {
+    int r = _read_bdev_label(cct, p, &label, true);
+    if (r < 0) {
+      if (r == -EIO) {
+        derr << __func__ << " fsck error: all block device labels corrupted" << dendl;
+      }
+      return r;
+    }
+  }
+  if (repair) {
+    string value;
+    int r = read_meta("replicated_label", &value);
+    if (r == 0 && value == "yes") {
+      int r = _replicate_bdev_label(p, label, false);
```

Comment:
Please clarify what label do we use here as a source for the recovery? Presumably the one which we managed to read successfully above. But what if there exist two different but both valid labels, e.g. after an incomplete previous replication? Which one to choose? IMO that's another design flaw. We really need epoch/timestamp/seqno in the label to distinguish the latest replica.